### ISSUE TITLE
Design: New Hashtrees

### ISSUE THREAD

**Issue by msteffen** [pachyderm/pachyderm#3116]
## Objective
Make it possible to delay merging hashtrees after a job has finished, so that the marginal time that a job takes reflects the marginal amount of data processed in the job (i.e. adding a small amount of data to a large commit should not take long to process)

## Design
- This idea is an extension of the representation described in https://github.com/pachyderm/pachyderm/issues/3048.
- The goal of a hashtree is the same: model a collection of key/value pairs, where keys are paths and values are NodeProtos, or slices of a block in object store

![new_hashtree](https://user-images.githubusercontent.com/235674/43972030-ed377bba-9c87-11e8-8905-cf0914e52424.png)

The diagram is described bottom-to-top of the highlighted, green box:
##### Blocks (containing file data)
- At the bottom of the green box there are PFS blocks containing file content, indicated as `[data――data]`.
  - These are the same as the blocks we already store in PFS
  - Slices of the block have been shaded to show where a hypothetical NodeProto might point
##### Value Blocks
- Above that are `value blocks`, indicated as `[N(blockref, br, ..., br)N(...)N(...)]`
- A `value block` is a block in object store containing a sequence of NodeProtos
- Each `N(...)` in the diagram above indicates a sequence of bytes of the form: `[length][serialized NodeProto]`, so the overall structure of a `value block` is `[length₁][NodeProto₁][length₂][NodeProto₂]...[lengthₙ][NodeProtoₙ]`
  - Each NodeProto contains a list of BlockRefs, where each BlockRef points to a region in a data block (indicated by the three arrows from a single NodeProto to block slices).
- There are no paths/keys present anywhere in a `value block`
##### Key Blocks
- At the top is a `key block`
  - `key block`s begin with a pointer to their parent hashtree (16 bytes)
    - The goal of this change is to avoid expensive merges for marginal jobs, so if a job only contains a few additional datums, it will now output a hashtree containing the marginal output and a pointer to its parent
    - Eventually a stack of hashtrees may be merged into a master hashtree. This design is based on LSM-trees and each hashtree is a single layer.
  - Next, `[bloom filter]` is a bloom filter containing the keys in this `key block`. This is so that if many key blocks are being queried simultaneously, irrelevant blocks can quickly be skipped.
  - Next, `[block code]` is a map `s --> block_id` where `block_id` is a `value block` and `s` is a small int (a simple compression mechanism; we refer to `s`es to avoid repeating the much longer `block_id`s)
    - Unclear if this should be `[total length][varint s][16-byte block_id][varint s][16-byte block_id]` or some other encoding
  - Then we have an ascending sequence of **k**ey, **b**lock **o**ffset (kbo) triples, ordered lexicographically by the key.
  - The sequence is partitioned into blocks (`u`) where each `u` is a fixed size (e.g. 1MB)
    - This makes it possible to read from the middle of a `key block` by reading at an offset
    `|parent ref| + |bloom filter| + |block code| + (N * 1MB)`
    - This in turn makes it possible to binary search the `u` blocks when looking up a given key.
    - To ensure each `u` is 1MB, we serialize keys into a key block and then, when a marginal key would cross a 1MB boundary, we instead write padding bytes to reach the boundary, e.g. with a sequence of `0`s, or some byte that can't start a key  (`[pad]` above)

This may look reduntant: why have "key+block+offset" tuples in the `key block` that point to *another* `block+offset`, in the NodeProto? Why not just store `[key][NodeProto]` together directly, as we do in https://github.com/pachyderm/pachyderm/issues/3048? The answer is that there is a crucial difference between the `key block`s above and `[key][NodeProto]...[k][N]`: `NodeProto`s can contain references to several blocks, but the block/offset data in a `key block` can only point to one block.

In practice, `NodeProtos` can be large if a file was written to by many jobs, so the block+offset representation above gives us a pointer to a potentially larger list of block pointers in the NodeProto

### Reading a File
To read a single key `k` from a commit `c`:
1. Trace the `parent ref`s of `c`'s hashtree all the way to the bottom of the stack (root hashtree)
2. For each Hashtree `h` in the stack:
    * a. if `k` is absent from `h`'s bloom filter, ignore `h` (but follow this process for the other hashtrees in the stack)
    * b. Otherwise, start by binary searching through the `u` blocks (remember `u[i]` can be read at a known offset) to find the block containing `k` (the largest initial key that is still < `k`)
    * c. scan that `u` block until `k` is found (or stop if not found)
    * d. Read the block+offset after `k` (pass through `block code`), and start reading the `value block` at that offset. Read an integer `[length]` followed by `length` bytes, and deserialize the bytes as a NodeProto
3. For all NodeProtos retrieved from some `h` in the stack, start reading bytes from the referenced blocks. Read bytes from the bottom hashtree's NodeProto, followed by the second-oldest NodeProto (which were logically appended to the first) all the way up to the top hashtree's NodeProto (the original hashtree referenced by `c`)

### Creation
A complication presented by this design is that we need to compute a hashtree's bloom filter before we can begin writing the hashtree's keys.
  - For input repos, we can keep a running bloom filter and update it after each call to PutFile (however, after a single call to DeleteFile, we need to recompute the bloom filter using the non-removed keys from the parent hashtree + the keys added in this commit)
  - When uploading a datum's output, we need to scan `/pfs/out` twice: Once to enumerate the file paths and compute the bloom filter, and then again to write the keys and values
  - Merging hashtrees is simpler, though—the component hashtrees' bloom filters can be ORed to create the output hashtrees bloom filter 

### Merge Algorithm
- First, note that with this representation, two hashtrees are already "merged" if the `parent ref` of one hashtree to point to the other.
- This is how this representation makes small jobs fast: by taking the full merge of output hashtrees out of the critical path of finishing a job.
- This comes at the cost of slower downstream reads, though, and eventually we will want to merge linked hashtrees into a single data structure, for fast reads.

To fully merge hashtrees *h<sub>1</sub>*, ..., *h<sub>n</sub>*:
1. Follow each *h<sub>i</sub>*'s `parent ref` to the bottom of the stack, and add all of those to the set *{h<sub>1</sub>*, ..., *h<sub>n</sub>}* of hashtrees to be merged
    - This won't be necessary for merging hashtrees at the end of a job, as datum hashtrees won't have any `parent ref`s
    - However, if a commit's hashtree is the top of a large stack, and we want to consolidate the stack for performance, we'll use this algorithm to do that. In this case, the only hashtrees being merged are the ones reachable by following `parent ref`s from the top hashtree
    - The output hashtree will have no parent ref
2. OR all *h<sub>i</sub>*'s bloom filters. The result is the output hashtree's bloom filter
3. Merge the block code by reading the *h<sub>i</sub>*'s block codes sequentially. If *h<sub>1</sub>*'s block code looks like:
  `1 -> 0x123, 2 -> 0x234, 3 -> 0x345`
  and *h<sub>2</sub>*'s block code looks like:
  `1 -> 0xabc, 2 -> 0xbcd, 3 -> 0xcde`
  Then the output hashtree's block code would start:
  `1 -> 0x123, 2 -> 0x234, 3 -> 0x345, 4(3+1) -> 0xabc, 5(3+2) -> 0xbcd, 6(3+3) -> 0xcde`
  `( +h₃: 7(6+1) -> ... )`
  All hashtrees' block codes can be concatenated like this: to add an additional hashtree `h'`, just add the output hashtree's largest `s` to every `s` in `h'`'s block code.
4. The `kbo` tuples can be merged using mergesort
    * a. select the smallest unmerged key in any *h<sub>i</sub>* and append it to the output hashtree's `kbo` tuples (if it's unique across *h<sub>i</sub>*). No need to read/parse the key's NodeProto!
    * b. If the lowest unmerged key is in multiple hashtrees, then read the NodeProto pointed to by all of those keys, merge them<sup>1</sup>, write the new NodeProto to a new `value block`, and write a `kbo` pointing into the new value block

Note that this merge algorithm only reads a small number of NodeProtos. Mostly output `kbo` tuples point to NodeProtos in the component hashtrees' `value block`s

<sup>1</sup> This part is underspecified. You need to merge parents before children, so you do have a slightly different merge process when flattening a stack vs. finishing a job. You could give each *h<sub>i</sub>* its own meta-key, and merge NodeProtos from low-meta-key *h<sub>i</sub>* before high-meta-key *h<sub>j</sub>*. This way, merging datums such that output from lexicographically early datums comes before lexicographically later datums, and flattening a stack of hashtrees where old file content comes before new file content, are handled the same ways

### Compression
The `u` blocks in the diagram above can be compressed, so that they hold more keys, as only the first key in each block needs to be read while binary searching `u` blocks (during point reads), and `u` blocks are only ever scanned from the beginning (which is amenable to streaming decompression with snappy)

#### B-Tree-like model
- This was more similar to the representation in https://github.com/pachyderm/pachyderm/issues/3048
- Basically, we would keep the hashtree representation `[path][NodeProto][path][NodeProto]...[path][NodeProto]` but we would also store `skip blocks` of the form `[path][offset][path][offset]...[path][offset]`, where the kth path in a `skip block` is the (k⋅N)th path in the overall hashtree, where N is e.g. 1e6, followed by the offset of the (k⋅N)th key in the hashtree
- To retrieve a key `k`, we'd scan the `skip block` to find the largest path less than `k`, jump to that path's offset in the hashtree, and then scan the hashtree until we found `k`
- To merge, we'd regenerate the `skip block` every time we wrote a new hashtree
- This representation is similar to a B-Tree because, in the case where the `skip block` got large, we could have a second-order `skip block` where its offsets pointed into the first-order `skip block` (like large nodes in a B-Tree)
- We abandoned this design due to the fact that it makes hashtrees uncompressable. IIUC, it's not possible to decompress a snappy-encoded byte stream from the middle, yet the point of the `skip block` is to allow us to start reading the hashtree from the middle

Other notes:
- Based on some cursory reading on my part, this seems most similar to the [representation that Cassandra uses](http://distributeddatastore.blogspot.com/2013/08/cassandra-sstable-storage-format.html) (Cassandra has a table of the form `[key][value][key][value]...[key][value]`, then a separate index into that table, then potentially a secondary index into the index (like a large B-tree fixed at two levels)

**Comment by msteffen** [comment]
This is similar to, but distinct from, the hashtree design we used for Pachyderm 1.8. Closing due to staleness.
